### DISCUSSION TITLE
Tabeller

### DISCUSSION THREAD

**Discussion by onordbo** [fremtind/jokul#2333] [Annet]
Som et ledd i [inventory prosjektet](https://github.com/fremtind/jokul/projects/6) vårt ønsker vi tilbakemelding og stemmer fra både design og teknologi for å kartlegge hvilke komponenter og behov vi burde prioritere. Nedenfor følger en kort hypotese/slutning fra Designsystem-teamet, i tillegg til en rekke kommentarer vi har merket oss gjennom arbeidet.

Se på skjermbildene fra inventory [tilgjengelig her for referanse](https://www.figma.com/file/66kOJ5nCHjgtkfLW2sQWB8/Jøkul-Inventory-Innsiktsyntese?node-id=0%3A1). Legg gjerne til ønsker, behov, tanker, meninger og annet som kommenaterer i denne tråden, så vi kan sørge for at oppgavene blir så godt beriket med informasjon som mulig, dette gjør det enklere for alle involverte å ta tak i og løse de utfordringene vi møter sammen.

***

### TL;DR 👉 “Ingen" bruker Jøkul-tabellen fordi den er veldig utdatert.

- Alle bruker en form for tabell, men ingen bruker Jøkul, og ingen deler kode
- Jøkul-tabellen er for snever, dekker ikke behovene
- Trenger definere når noe er en tabell og når noe bare ser litt sånn ut, men er noe annet
- Tabeller vises i flere contexter. Helside, modal, accordion, kort.
- Ulik kompleksitet ute i løsningene; alt fra ren datatabell uten interaktivitet til trenivå nøstet ekspanderbar tabell med handlinger
- Ekspanderbar rad går igjen i flere løsninger, med varierende innhold
- Varierer måte å vise ekspanderbar på: chevron på høyre vs “Accordion-pil” venstre
- Vi må ha en mobilvennlig visning av tabellinnhold. BM løser ved å gjøre om rader til kort. Andre tabeller må scrolle horisontalt.
- Tabeller brukes i flere ulike sammenhenger: helside, modal, accordion
- Internsystemer for superbrukere trenger mer data og mindre luft
- Trenger guidelines for interaktive rader. Skal hel rad være klikkbar vs kun knapper. Husk touchflate på mobil.
- Ulik måte å vise interaktivitet på (hover state): “scale”, bakgrunnsfarge, fontfarge
- Ulik typografi i tabellheader. Noen større enn innhold, noen mindre, noen lik str. men bold
- Tabellheader noen ganger “sticky” ved scrolling, som oftest ikke
- Noen tabeller tvunget til paginering pga ytelse. Mangler Jøkul-komponent, prinsipper.
- Flere tilfeller av filtrering i tabell. Skjemafelter vist over tabellheader

**Comment by wkillerud** [comment]
Finnes en del issues på tabeller opp igjennom som kan være greit å ha med i diskusjonen videre:

- #841
- #951
- #1341
- #1415

Vil spesielt fremheve #951. Ikke gitt at valgene som ble tatt den gang fremdeles skal gjelde, men likevel. Den siste går på å ikke bruke `box-shadow` til borders fordi de forsvinner i Safari.

**Comment by karimarkhus** [comment]
### Oppsummering fra møte om tabeller med designer og utviklere fra BM, SAM og Flyt + Øyvind og William fra designsystemet. 
 
Vi gikk gjennom de ulike løsningene og ble enige om å samarbeide videre på design og utvikling i de forskjellige teamene. Mye er allerede utviklet, og man kan muligens bruke en del av den koden som fins allerede.
 
**Felles behov:**
- Klikkbare rader: her gjør vi det litt forskjellige i dag, noen bruker knapp, mens andre har hover farge på hele raden for å vise at det er klikkbart
- Ekspbanderbare rader i ulike variasjoner
- Kunne se om noe har vært åpnet før
- Forhåndsvise innhold (en idé)
- Mobil versjon
- Ulik behov for mengde data, vi trenger derfor både vanlig og kompakt versjon av tabellkomponenten
- Bør definere labels. 
- Tabell vs accordion, hvordan bør de skille seg fra hverandre
- Mulighet for å scanne tabeller med mye data. Mulighet for “sebrastriper” ved behov?
- Light mode / dark mode
 
Vi ble enig om å samarbeide om en felles tabellkomponent, som man f.eks kan fylle med ulike ting, sånn som f.eks man kan gjøre med accordion. Lage byggekolosser.
Må definere noen regler, noen steder er det f.eks et behov å ha flere klikkbare felt på samme rad. 
 
**Videre plan:** 
- Hva er worst case scenario: Lage en skisse på mest komplekse tabell, og lage definisjoner på dette eksempelet
- Løse det med enkeltkomponenterRad -ekspanderbar - klikkbar - tabellrad handling - Dokumentert begge ender av skalaen
- Samarbeid mellom (Jo Emil) BM og (Simen/lars Are) Flyt på utvikling med backing fra William (designsystemet)

**Design:** 
- skisser opp en worst case tabell
- Tar en kjapp status på design og noen prinsipper: Spacing, pila, typografi, active/hover, mobil, dark mode, klikkbare rader/eller knapp, Solomodus-flere åpne accordions ad gangen, Scanning.

**Utvikling:**
- Komponentinndeling, Grovsortering, Inventory - William holder i det? Samarbeider med (Jo Emil) BM og (Simen/lars Are) Flyt 
 
 **Felles:**
- Vi oppdaterer hverandre og møtes når design har en skisse på worst-case tabell.

**Comment by onordbo** [comment]
## Oppdatering Designsystem-Forum (via VD-forum):
### Generell status: Visuell utforming er fortsatt under arbeid. Teknisk utredning og grunnleggende implementering basert på inspirasjon fra Flyt er påbegynt.
- Størrelse og spacing ivarets som Stein Olav gjorde for en stund siden.
- Vi ser nærmere på ekspandering da det er noe divergering mellom BM og Flyt.
- Bruk av pil varierer mellom løsninger, vi må enn så lenge tenke litt mer på det. Det kan skape en del visuell støy, vi burde se pil i sammenheng med kontekstuelle behov
- Vi vurderer fortsatt konteksten for bruk av chevron, og dennes forhold til en tabell som også har pil
- Vi har diskutert om vi skal fremheve labels eller ikke, men vi ser at vi har ganske ulike behov her, det kommer an på innholdet i tabellen, noen ganger er det veldig nødvendig, andre ganger ikke. Vi ser for oss at vi har to varianter, hvor man kan velge mellom en tydligere og en mer nedtonet versjon.
- Det er lurt at vi gjør det enkelt rent teknisk i starten, så kan vi gjøre det mer komplekst etterhvert

### Hva skjer nå?
- [ ] Vi jobber videre med ekspandering på designsiden (@Murstam)
- [ ] Utrede hoverstates (@Murstam)
- [ ] Sticky-labels og nøstede tabeller må utforskes mer (@Murstam)
- [ ] Vi begynner å bygge en grunnleggende tabell slik at vi kan få litt fart på en basiskomponent (@wkillerud)

**Comment by joms** [comment]
Jeg har fått tilbakemelding fra enkelte på at de små og mindre synlige headerene over kolonnene blir litt borte. Usikker på om dette bare bryter med at man er vant til at headerene er mest synlige, eller om det faktisk gjør det vanskeligere å oppdage headerene.

**Comment by wkillerud** [comment]
Utviklingen er i gang. En sneak peek på APIet: #2380

**Reply by ghost** [reply]
@wkillerud  Nydelig. Dette gjør at man kan bygge tabellen slik man vil ut fra manges behov, men får samtidig ett unified design. Hvis man nå vil ha f.eks en strikethrough på tekst inne i en TableCell er nå dette mulig. Jeg foreslår to nye komponenter også: TableRowSpacer evt. TableRowRuler siden man i visse tilfeller har lyst til å benytte en hel tabellrad for luft eller en hel tabellrad ruler for å skille mellom innhold inne i samme tabell. Da unngår man at flere benytter forskjellig farge, custom CSS med varierende høyde eller prøver å fylle ut en rad med tomt innhold etc, for å skille mellom innhold i samme tabell hvis de ønsker det:

Eksempel:
![spacer](https://user-images.githubusercontent.com/92525629/137329329-a895d11b-ba6c-4368-8bac-f492975277a3.png)

**Reply by wkillerud** [reply]
Takk for tilbakemelding @e-sving!

Skjermbildet ditt satte i gang en diskusjon her i landskapet om "hva er en tabell og hva er en liste (av tabeller)" 😄 Det hadde vært interessant å grave litt dypere i caset ditt på Designsystemforum!

**Reply by ghost** [reply]
Skjønner det er ett tvilstilfelle:)

Hvis det er 2 tabeller eller en liste med tabeller, en over og nedenfor ruler så regner jeg med at kolonnene i Tabell1 og Tabell2 ikke vil ha riktig horizontal alignment hvis teksten i en TableCell er mye lenger i den nederste tabellen (eller den øverste) eller hvis det er veldig varierende dynamisk innhold i de to tabellene, men det igjen kan jo fikses med å sette en fixed width på kolonnene i en tabell.

Ble usikker selv her, dere får diskutere litt om det er ett behov for spacere.

**Comment by wkillerud** [comment]
Da begynner førsteversjonen av tabellkomponentene å ta form. Tar gjerne en gjennomgang av designet, kanskje spesielt på mobil hvor jeg måtte freestyle litt. ~~Ikke veldig fornøyd med mobilvisningen av ekspanderende listevariant.~~ (ekspanderbar ikke lenger relevant)

Et skjermbilde fra dokumentasjonen med eksempler:

![localhost_8000_komponenter_table (1)](https://user-images.githubusercontent.com/1223410/137732020-c3961242-74c8-46b3-bc31-1dd060d8363d.png)

Og en video av interaktiviteten:

https://user-images.githubusercontent.com/1223410/137732590-b1e75992-bedf-483b-91ef-7645f905571b.mov

Cc @Murstam

**Reply by wkillerud** [reply]
Oppdatert stablet mobilvisning:

![image](https://user-images.githubusercontent.com/1223410/137750980-6e415da9-b731-416b-ac03-7cade647a081.png)

**Comment by wkillerud** [comment]
Etter diskusjon i pull requesten #2380 utsetter vi å inkludere ekspanderbar-funksjonaliteten fra Jøkul til vi har tid til å dykke enda dypere inn i materien når vi skal se på patterns. Det sagt så er det mulig å bruke Jøkul-komponentene for å lage sin egen ekspanderbare rad ute i applikasjonene inntil videre.

**Comment by wkillerud** [comment]
Da har Øyvind og jeg tatt en kjapp design-QA og sier oss fornøyde nok til å gå live med tabellene slik de er nå. Tenker holde pull requesten åpen for kommentarer ut uka, så blir det å publisere dem på mandag.

**Comment by MonaHolm72** [comment]
I gitte situasjoner hvor en verdi i en rad har tekst som ikke bryter over to linjer, burde teksten toppjusteres i stedet for midstillingen. Dette skaper en ujevn og rotete leseropplevelse i tabeller.

<img width="1146" alt="tabell" src="https://user-images.githubusercontent.com/66421429/139407438-68042d6f-cb33-4f1e-8edc-071fec304de4.png">
.

**Reply by wkillerud** [reply]
Vi har en variant for å toppjustere innholdet for de som trenger det, men det er litt vanskelig å helgardere seg mot effekten på alle skjermstørrelser.

**Reply by onordbo** [reply]
Enig i at det blir vanskelig å helgardere seg. Hva tror du om å omdefinere standardinnstillingen fra midtstilling til toppjustering, @wkillerud?

**Reply by wkillerud** [reply]
Tenker vi bør la det gå litt tid før vi tar den beslutningen. Om vi ser at "alle" gjør det kan vi endre default ganske lett.

**Reply by MonaHolm72** [reply]
Midtstilt kan ikke være default...hopp og dans. Men jeg korrigerer meg selv til bunnjustert. Som er normen i tabelloppsett og mest leservennlig. o-ha.

<img width="765" alt="tabell 2" src="https://user-images.githubusercontent.com/66421429/139449476-0d816fd4-7791-4663-927b-0b8996f34798.png">

**Reply by karimarkhus** [reply]
@MonaHolm72 @onordbo @wkillerud syns også at det bør justeres til toppen, spesielt med tanke på at det noen ganger er mer innhold enn bare en rad, eller flere rader som er linket mot hverandre. tror lesbarheten kan bli ganske rotete.

**Reply by onordbo** [reply]
Ja, jeg tenker også at det kan oppleves som ganske rotete, spesielt i FLYT (eller alle steder det er _mange_ rader). Har du en mulighet til å lage en røff testskisse av et case fra Flyt som viser forskjellen mellom midtjustert og toppjusterte rader @karimarkhus? 😄

**Reply by karimarkhus** [reply]
![Screenshot 2021-11-01 at 12 36 02](https://user-images.githubusercontent.com/91268511/139665782-a00011ae-3b5e-47c1-b8e1-fc5c81c4123a.png)
![Screenshot 2021-11-01 at 12 35 04](https://user-images.githubusercontent.com/91268511/139665787-27b22f38-c317-4d48-9237-8d2a348b9eec.png)

her er 2 eksempler @onordbo

**Reply by onordbo** [reply]
Ja her ser man jo ganske fort hva som funker bra og hva som ikke funker så bra—jeg holder definitivt en liten nål på at vi burde legge oss på toppjustering som standarden. (cc @MonaHolm72 @wkillerud)

**Reply by MonaHolm72** [reply]
Topp er topp!

**Reply by karimarkhus** [reply]
toppers!

**Comment by Stakset** [comment]
Jeg synes vi skal prøve å følge samme strukturen på alle sidene i Jøkul portalen. Akkurat nå skiller siden til tabeller seg en del fra de andre komponentene. Mye tekst før man kommer ned til selve komponenten. Og kunne man ikke lagt flere av variantene i samme eksempel ved å bruke egenskaper? Type: "tabelloverskrifter, listevisning, handling.."

@onordbo @wkillerud @Slaine2

**Reply by wkillerud** [reply]
@piofinn etterlyste også et synlig eksempel uten å scrolle så mye, så det er på vei i #2485 

Jeg er litt mer usikker på å samle eksemplene. Er redd for at de blir for adskilt fra beskrivelsene av hvordan de skal brukes. Men på den andre siden er de kanskje vanskeligere å oppdage i utgangspunktet sånn de er nå?

**Reply by onordbo** [reply]
Jeg tenker ja takk begge deler @wkillerud. Et godt oppsummert eksempel i toppen i tillegg til mer spesialiserte eksempler for nisjesituasjoner lenger nede. Hovedmålet _(slik jeg ser det)_ for toppen av siden til en komponent burde vel være å gi et et "executive summary" raskt og enkelt.

Eventuelt hvis vi ser at en side blir **veldig lang**, kanskje legge til noen anker-snarveier i toppen?

**Comment by Stakset** [comment]
Kan jeg legge inn et ønske om at overskriftene har en mulighet for "regular"? Ikke alle tilfeller hvor man ønsker å sette fokus på de. @wkillerud

**Reply by wkillerud** [reply]
Det er teknisk sett mulig allerede. Vi har en `bold`-property som kan settes til `false` i koden. Ser det ikke er så godt dokumentert utover at "[om man] er usikker, bruk en kompakt og bold overskrift" som hinter om at det finnes en "ikke-bold". Tanken var vel at en skulle kunne ha noe frihet til å tilpasse til innholdet rundt (cc @Murstam).

**Reply by Stakset** [reply]
Takk for svar. Da ordna det seg :)

**Comment by onordbo** [comment]
https://github.com/fremtind/jokul/issues/2392
